platform/posix/transport/src/openssl_posix.c

```diff
@@ -0,0 +1,674 @@
+        sslStatus = setClientCertificate( pSslSetup,
+                                          pOpensslCredentials->pClientCertPath,
+                                          pOpensslCredentials->clientCertPathLen );
+    }
+
+    if( ( sslStatus == 1 ) &&
+        ( pOpensslCredentials->pPrivateKeyPath != NULL ) &&
+        ( pOpensslCredentials->privateKeyPathLen > 0 ) )
+    {
+        sslStatus = setPrivateKey( pSslSetup,
+                                   pOpensslCredentials->pPrivateKeyPath,
+                                   pOpensslCredentials->privateKeyPathLen );
+    }
+
+    return sslStatus;
+}
+
+static void setOptionalConfigurations( SSL * pSslContext,
+                                       const OpensslCredentials_t * pOpensslCredentials )
+{
+    int sslStatus = -1;
+    char * sniHostName = NULL;
+
+    /* Set TLS ALPN if requested. */
+    if( ( pOpensslCredentials->pAlpnProtos != NULL ) &&
+        ( pOpensslCredentials->alpnProtosLen > 0 ) )
+    {
+        LogDebug( ( "Setting ALPN protos." ) );
+        sslStatus = SSL_set_alpn_protos( pSslContext,
+                                         ( unsigned char * ) pOpensslCredentials->pAlpnProtos,
+                                         ( unsigned int ) pOpensslCredentials->alpnProtosLen );
+
+        if( sslStatus != 0 )
+        {
+            LogError( ( "SSL_set_alpn_protos failed to set ALPN protos. %s",
+                        pOpensslCredentials->pAlpnProtos ) );
+        }
+    }
+
+    /* Set TLS MFLN if requested. */
+    if( pOpensslCredentials->maxFragmentLength > 0 )
+    {
+        LogDebug( ( "Setting max send fragment length %lu.",
+                    ( unsigned long ) pOpensslCredentials->maxFragmentLength ) );
+
+        /* Set the maximum send fragment length. */
+        sslStatus = SSL_set_max_send_fragment( pSslContext,
+                                               ( long ) pOpensslCredentials->maxFragmentLength );
+
+        if( sslStatus != 1 )
+        {
+            LogError( ( "Failed to set max send fragment length %lu.",
+                        ( unsigned long ) pOpensslCredentials->maxFragmentLength ) );
+        }
+        else
+        {
+            /* Change the size of the read buffer to match the
+             * maximum fragment length + some extra bytes for overhead. */
+            SSL_set_default_read_buffer_len( pSslContext,
+                                             pOpensslCredentials->maxFragmentLength +
+                                             SSL3_RT_MAX_ENCRYPTED_OVERHEAD );
+        }
+    }
+
+    /* Enable SNI if requested. */
+    if( ( pOpensslCredentials->sniHostName != NULL ) &&
+        ( pOpensslCredentials->sniHostNameLen > 0 ) )
+    {
+        LogDebug( ( "Setting server name %.*s for SNI.",
+                    ( int32_t ) pOpensslCredentials->sniHostNameLen,
+                    pOpensslCredentials->sniHostName ) );
+
+        /* NULL-terminated string is required for SSL_set_tlsext_host_name. */
```

Comment:
For all these scenarios requiring a malloc for NULL termination, is it better to ask the caller to provide NULL terminated strings?

Reply:
I am doing this more for convention's sake as we are avoiding the use of NULL-terminated strings.